### ISSUE TITLE
Topic-encryption: Shared DEKs between multiple threads/channels

### ISSUE THREAD

**Issue by robobario** [kroxylicious/kroxylicious#775]
We want to have control over the memory consumption of DEKs and make efficient usage of the DEKs we do generate. If we can encrypt the data using fewer DEKs then there will be less key material to retrieve at decrypt time.

Currently we maintain a cache of DEKs and their associated ciphers/encrypter classes per channel. So this could cost a lot if we eventually have to support thousands of concurrent connections. On the upside the blast radius of a compromised DEK is smaller, only exposing the data for one client.

There are several schemes we could use to share the DEKs:

1. thread local DEK caches, so each eventloop has a DEK cache. Could be efficient from the locking point-of-view if we could keep all the DEK encrypt work on that eventloop. So multiple channels on the same eventloop would be able to use the same single-threaded DEK encryption classes without blocking.
2. global DEK cache, with associated co-ordination if we are trying to share components like ciphers/encrypters to ensure we only use a DEK up to its maximum encrypt count and use non-threadsafe encrypters appropriately.
3. per-virtual-cluster cache. similar to the global case but bound to a VC so we don't share DEKs across VCs

There is likely more schemes you could devise. Like the key material could be cached globally and then be used to create multiple single-threaded ciphers/encrypters that could be pooled or local to an eventloop (taking care to co-ordinate it so that we never exceed the encryptions per DEK).

**Comment by robobario** [comment]
@tombentley do you think this ticket could be closed with the decryption path change? Now we have a per-filter-configuration cache reusing DEKs. The caches don't share DEKs but we're no longer operating per-channel.

**Comment by tombentley** [comment]
Yes, I think this can be closed. I'll open another issue for the possibility of passing freshly minted DEKs to the decryption path, but that one doesn't need to be in 0.5.0

**Comment by robobario** [comment]
Summary of the Implementation:

1. We introduced a `DekManager` and `Dek` abstraction to encapsulate and protect the secret key material of the data encryption keys. This is to prevent our client code getting a reference to a secret key and using it willy-nilly for unlimited cryptor operations. We must limit how much work a DEK does. This also included the bones for sharing the DEK across channels using reference counting to control when we destroy a DEK. #901, #893 
2. We introduced a cache from (Kek, CipherId) -> DEK on the encrypt path. This cache is created once when initialising the filter and is shared across channels. So all channels the create an instance of this Filter can re-use that DEK while it has cryptor operations remaining. The stateful, non-threadsafe Encryptors obtained from the DEK are only used to encrypt a single topic-partition's MemoryRecords before they are closed. #1001
2. We introduced a cache from (EDEK, CipherId) -> DEK on the decrypt path. This cache is created once when initialising the filter and is shared across channels. So all channels the create an instance of this Filter can re-use that DEK for decryption. The stateful, non-threadsafe Decryptors obtained from the DEK are only used to decrypt a single topic-partition before they are closed. #1002
